feat: make it possible to restore default highlighting

habamax opened the pull request:
problem: users are unable to restore default highlighting
solution: save and restore default links

prev solution: treat cleared highlight group as "has no hl-settings"

Fixes #4405

habamax committed:
feat: make it possible to restore default highlighting

problem: users are unable to restore default highlighting
solution: treat cleared highlight group as "has no hl-settings"

Fixes #4405

brammool commented:
I'm wondering: would it also work to remove the SG_LINK flag from sg_set in highlight_clear() ?

Since you know what you are fixing, can you also add a test for that?

habamax committed:
feat: add Test_highlight_restore_defaults

`highlight default` should work for 'cleared' groups.

habamax commented:
> Since you know what you are fixing, can you also add a test for that?

Test has been added.


> I'm wondering: would it also work to remove the SG_LINK flag from sg_set in highlight_clear() ?

Looks like it also solves the issue.

Removed check for `cleared` in `hl_has_settings`

Added `HL_TABLE()[idx].sg_set = 0;` to `highlight_clear`


Should I make this change instead?

habamax commented:
> Should I make this change instead?

It looks like this flag is reset outside of the `highlight_clear`:

https://github.com/vim/vim/blob/8b51b7f0f17af149a8ce76e805050977857f9e50/src/highlight.c#L882-L888

But I am not sure if there is a combination of `forceit==1` and `init==1` and `doclear == 0` we can call from vimscript, as was revealed by @andymass: https://github.com/vim/vim/issues/4405#issuecomment-494598219

habamax committed:
style: fix indentation

habamax committed:
feat: save and restore default links

This approach is from Antony Scriven @adscriven:

* default hi-links are saved on `hi def`
* and restored in `hi clear`

Having this user doesn't have to do neither `syntax on` nor `syntax
enable` to be able to apply default links on a cleared highlight groups.

Thus there will be no need to update old colorschemes to make it work.

habamax commented:
Bram, I have actually just pushed a new variant, like a seconds before you closed the PR :)

It might be better in the end, but it is up to you to decide.

> This approach is from Antony Scriven @adscriven:
> 
> * default hi-links are saved on `hi def`
> * and restored in `hi clear`
> 
> Having this user doesn't have to do neither `syntax on` nor `syntax
> enable` to be able to apply default links on a cleared highlight groups.
> 
> Thus there will be no need to update old colorschemes to make it work.

andymass commented:
@habamax does Bram's commit resolve the issue I pointed out in that comment?  (If so can you briefly explain how?)

habamax commented:
@andymass there is a code path checking if default link is being set for a highlight group and if it has any settings. 

```
    // Return if "default" was used and the group already has settings.
    if (dodefault && hl_has_settings(idx, TRUE))
       return;
```

That function was always false for a cleared highlight group thus preventing :hi def GroupName

So with the patch cleared hi group is treated as the group without hl-settings.

brammool commented:
> Bram, I have actually just pushed a new variant, like a seconds before you closed the PR :)
> 
> It might be better in the end, but it is up to you to decide.

This would restore default links?  Is that actually needed?

	% vim --clean
	:hi
	" Shows Character      xxx links to Constant
	:hi link Character Number
	:hi
	" Shows Character      xxx links to Number
	:hi clear
	:hi
	" Shows Character      xxx links to Constant


-- 
Did Adam and Eve have navels?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

adscriven commented:
On Sep 16, Bram Moolenaar wrote:

 > [...]
 >
 > This would restore default links?  Is that actually needed?
 >
 > [...]

E.g. syntax/html.vim links htmlTagName to Statement, but a fancy
colorscheme may choose to give htmlTagName specific colours. This sort
of thing happens frequently.

If you then switch from one of these fancy colorschemes to one that uses
only the standard groups, htmlTagName won't be highlighted with
Statement colours as you might expect, because the link has gone. It
looks like the second colorscheme is broken.

The idea was that if default links as well as default colours are
restored on ":hi clear", then colorschemes will be working with a more
consistent initial state, and produce the colours that the user expects
to see.

--Antony

brammool commented:
> On Sep 16, Bram Moolenaar wrote:
> 
>  > [...]
>  >
>  > This would restore default links?  Is that actually needed?
>  >
>  > [...]
> 
> E.g. syntax/html.vim links htmlTagName to Statement, but a fancy
> colorscheme may choose to give htmlTagName specific colours. This sort
> of thing happens frequently.
> 
> If you then switch from one of these fancy colorschemes to one that uses
> only the standard groups, htmlTagName won't be highlighted with
> Statement colours as you might expect, because the link has gone. It
> looks like the second colorscheme is broken.
> 
> The idea was that if default links as well as default colours are
> restored on ":hi clear", then colorschemes will be working with a more
> consistent initial state, and produce the colours that the user expects
> to see.

Can you make a new pull request, so we can see the difference with the
already included solution?  The test would also need updating.

-- 
The Characters and incidents portrayed and the names used are fictitious and
any similarity to the names, characters, or history of any person is entirely
accidental and unintentional.
                                  Signed RICHARD M. NIXON
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

habamax commented:
It is in an old PR as a new commit, I can make it as a separate PR.



ср, 16 сент. 2020 г., 22:50 Bram Moolenaar <notifications@github.com>:

>
> > On Sep 16, Bram Moolenaar wrote:
> >
> > > [...]
> > >
> > > This would restore default links? Is that actually needed?
> > >
> > > [...]
> >
> > E.g. syntax/html.vim links htmlTagName to Statement, but a fancy
> > colorscheme may choose to give htmlTagName specific colours. This sort
> > of thing happens frequently.
> >
> > If you then switch from one of these fancy colorschemes to one that uses
> > only the standard groups, htmlTagName won't be highlighted with
> > Statement colours as you might expect, because the link has gone. It
> > looks like the second colorscheme is broken.
> >
> > The idea was that if default links as well as default colours are
> > restored on ":hi clear", then colorschemes will be working with a more
> > consistent initial state, and produce the colours that the user expects
> > to see.
>
> Can you make a new pull request, so we can see the difference with the
> already included solution? The test would also need updating.
>
> --
> The Characters and incidents portrayed and the names used are fictitious
> and
> any similarity to the names, characters, or history of any person is
> entirely
> accidental and unintentional.
> Signed RICHARD M. NIXON
> "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD
>
> /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net \\\
> /// sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
> \\\ an exciting new programming language -- http://www.Zimbu.org ///
> \\\ help me help AIDS victims -- http://ICCF-Holland.org ///
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vim/vim/pull/6956#issuecomment-693629135>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZKFXGDL23NNLGFWYKTG3SGEJI3ANCNFSM4RMVTWQA>
> .
>

habamax commented:
> 
> 
> after look over #6970, maybe a silly question but one concern:
> if one colorscheme was on, and it set or override one 'hi def', then now issued 'colo default' (had a 'hi clear'), what would happened on that 'hi def', or just supposed that/all 'hi def' would not impact (not existed in) 'colo default' ?

if there was no syntax loaded, meaning there was no `:hi def GroupName` defined, then colorscheme will define it and `:hi clear` will make it `cleared`:

Colorscheme redefines
:hi vimOption ctermbg=red

Then you do:
:hi clear
:hi vimOption --> cleared

Before the patch, it would stay cleared and it was impossible to make it default again as `:hi def vimOption` just didn't work for cleared. You could only override it with `:hi vimOption` in some other colorscheme.

> specially 'hi clear', perhaps i would run 'hi clear' and/or specially some vim scripts had 'hi clear' in it, but user did not expect those such kind 'hi def' impact its previous logic/effect.
> was it possible to happen, or i missunderstood sth ?
> means just wish really only the default 'hi def' which not 3rd-party made ones to be restored after 'hi clear'.

With the #6970 patch vim also restores default links defined in syntax files, thus if some colorscheme redefines vimOption it would be restored on `:hi clear` as if it was initially defined in syntax file.

If you check issue description it has all the details: https://github.com/vim/vim/issues/4405

habamax commented:
> actually my concern was - for example:
> one 3rd part colorschema had a 'hi! def link difftext foo' in it, then now, after your patch, whatever 'colo default' or 'hi clear', after one of them, what would happened, would it back to original 'difftext' ? or still is 'foo' ? since now that 'def link' is 'foo' now.
> just wish you gave me a confirm answer. :)

with patches:

```
hi difftext
output: 
    DiffText       xxx term=reverse cterm=bold ctermbg=9 gui=bold guibg=Red

hi! def link difftext foo

hi difftext
output:
    DiffText       xxx term=reverse cterm=bold ctermbg=9 gui=bold guibg=Red
                   links to foo

colo default

hi difftext
output:
    DiffText       xxx term=reverse cterm=bold ctermbg=9 gui=bold guibg=Red
```
